lib/btree.js

```diff
@@ -0,0 +1,134 @@
+'use strict';
+
+const DEFAULT_DEGREE = 6; // min degree of b-tree.
+// All vertices except the root have [degree ... 2 * degree] child nodes
+// And [degree - 1 ... 2 * degree - 1] + 1 "empty" elements
+
+class Element {
+  constructor(value, tuple, child = null) {
+    this.value = value;
+    this.tuple = tuple;
+    this.child = child;
+  }
+}
+
+const empty = child => new Element(undefined, null, child);
+
+const destroyNode = (parent, index) => {
+  const node = parent[index].child;
+  const len = node.length;
+  const newLeftNode = node.splice(0, len / 2 - 1); // First half
+  const mid = node.shift();
+  newLeftNode.push(empty(mid.child));
+  mid.child = newLeftNode;
+  parent.splice(index, 0, mid);
+  return parent;
+};
+
+const isLeaf = node => !node[0].child;
+
+const binarySearch = (node, value) => {
+  let start = 0;
+  let end = node.length - 1;
+  while (start <= end) {
+    const i = (start + end) >> 1;
+    const item = node[i];
+    const itemValue = item.value;
+    if (value > itemValue) start = i + 1;
+    else if (!(value >= itemValue)) end = i - 1;
+    // undefined must be bigger then other value
+    // (!(1 >= undefined)) -> true, but just (1 < undefined) -> false
+    else return [true, i];
+  }
+  return [false, start];
+};
+
+function* inorderTraversal(start, finish, currNode) {
+  let [found, startIndex] = binarySearch(currNode, start);
+  if (found) startIndex++;
+  let finishIndex;
+  if (finish === undefined) {
+    finishIndex = currNode.length - 1;
+  } else {
+    let _; // We need to initialize the unnecessary variable to use
+    // the destructive assignment in the next line
+    [_, finishIndex] = binarySearch(currNode, finish);
+  }
+  for (let currIndex = startIndex; currIndex <= finishIndex; currIndex++) {
```

Comment:
What do you think of
```
if (isLeaf(currNode)) {
  for (let currIndex = startIndex; currIndex < finishIndex; currIndex++) {  
    yield currNode[currIndex].tuple;   
  }
} else {
  for (let currIndex = startIndex; currIndex < finishIndex; currIndex++) {
    const currElement = currNode[currIndex];
    yield* inorderTraversal(start, finish, currElement.child);  
    yield currElement.tuple;   
  }
  yield* inorderTraversal(start, finish, currNode[finishIndex].child);  
}
```
?

Reply:
Yes, I also wanted to write this way, and I thought that it would faster. But I was afraid that there is a copy-paste. If this is not very critical, I implement this variant